### ISSUE TITLE
Make "data" member of "relationships" optional

### ISSUE THREAD

**Issue by bduisenov** [yahoo/elide#285]
This is a follow-up enhancement to #282 . 
After adding `links` attribute to `relationships`, `data` attribute should become optional. 

@DeathByTape, @clayreimann proposal from my side is to modify `PersistentResource#getRelationshipsWithRelationshipFunction` to filter out `null` `relationships` retrieved from `relationshipFunction`, so that when `DataStoreTransaction` returns `null` it would not be included.

**Comment by clayreimann** [comment]
@bduisenov I disagree for two reasons. First, filtering `null`, but not `[]` create an inconsistency between to-one and to-many relationship types. Second–and more importantly–I think there is value in letting people know that this relationship is empty (and less compliant clients fail spectacularly if the `data` member is not present).

That being said, if you feel that there is a strong reason to not pass `data` if it is empty the I suppose we can hide this behind a builder flag so that when constructing and Elide service authors can decide if they want this behavior.

**Comment by DeathByTape** [comment]
Sorry for the delay @bduisenov.

Let me make sure I understand correctly (I started writing a response and then realized I misunderstood 😄), you want to ignore any `null` members and perhaps even `[]` (to comply with @clayreimann's comment) for relationships?

I don't think it's appropriate to hide the `relationships` field if there exist relationships on the object (regardless of their values). Since the [jsonapi spec](http://jsonapi.org/format/#document-resource-objects) states [MAY](https://www.ietf.org/rfc/rfc2119.txt), this is truly an optional field but I think we should present it when we have anything meaningful to display. What I mean is that I would be happy to hide the following:

```json
{
  ...
  "relationships": {}
  ...
}
```

but would be very hesitant to suggest that we omit something like this, however:

```json
{
  ...
  "relationships": {
    "myToOneRelationship": null
  }
  ...
}
```

Since that is actually information about the status of an existing relationship that the user has access to. In general, it's probably better for the server to be more explicit on these things than encoding this data implicitly on whether or not it exists. This could become even more problematic for any breaking-changes to the schema (assuming/relying on values that no longer exist, etc.).

My 2cents.

**Comment by bduisenov** [comment]
Hi @DeathByTape, @clayreimann.

Thanks for your replies.

What I want to have, is an option in Elide, not to even try to fetch relationships data upfront and to give a client only a link instead.

f.ex. with introduction of #282 the response would be:
```javascript
{
  "relationships": {
     "author": {
      "links": {
        "self": "/articles/1/relationships/author",
        "related": "/articles/1/author"
      },
      "data": { "type": "people", "id": "9" }
    }
  }
}
```

But I want:
```javascript
{
  "relationships": {
     "author": {
      "links": {
        "self": "/articles/1/relationships/author",
        "related": "/articles/1/author"
      }
    }
  }
}
```
instead.

My first idea was to override DataStoreTransaction#getRelation to return either null or empty collection or whatsoever to indicate that data should not be included into relationships. But concept is broken and there should be another, better way how to implement this. Do you have an idea?

Cheers,
Babur

**Comment by bduisenov** [comment]
Hi @DeathByTape,

`The goal for this improvement is not performance, is it?` - yes, it is.

Let's assume I have a complex system where `DTO`'s are constructed by calling multiple `populator`s/`converter`s. For example: `ProductDTO` might be constructed by calling Price/Images/Description/Classification/Url/Reviews/Promotions/etc. populators. Now, I want to build JSONAPI having `Facade`s with this `DTO`'s. 

I've added `JPA` annotations to `DTO`'s (ugly) together with `Elide` annotations. Wrote `delegator`s`/adaptor`s for `DataStoreTransaction` and `DataStore` to call this` Facade`s. When I call `/products/1` it returns me a product, all good. But, I also have `CategoryDTO` which has `toMany` `relationship` with products. In this case, when I call `/category/1` the `data` attribute would be included into `relationships` containing id's of all 5000+ products 😄.  I don't want to load `ProductDTO`s just for showing their `id`s in this case, instead showing a relationship link would be perfect for me.
